modules/xfeatures2d/src/daisy.cpp

```diff
@@ -0,0 +1,2057 @@
+
+      return out;
+    }
+
+    // computes the l2norm of an array: [ sum_i( [a(i)]^2 ) ]^0.5
+    template<class T> inline
+    float l2norm( T* a, int sz)
+    {
+      float norm=0;
+      for( int k=0; k<sz; k++ )
+         norm += a[k]*a[k];
+      return sqrt(norm);
+    }
+
+    // computes the l2norm of the difference of two arrays: [ sum_i( [a(i)-b(i)]^2 ) ]^0.5
+    template<class T1, class T2> inline
+    float l2norm( T1* a, T2* b, int sz)
+    {
+      float norm=0;
+      for( int i=0; i<sz; i++ )
+      {
+         norm += square( (float)a[i] - (float)b[i] );
+      }
+      norm = sqrt( norm );
+
+      return norm;
+    }
+
+    template<class T> inline
+    float l2norm( T y0, T x0, T y1, T x1 )
+    {
+      float d0 = (float) (x0 - x1);
+      float d1 = (float) (y0 - y1);
+
+      return sqrt( d0*d0 + d1*d1 );
+    }
+
+    // allocates a memory of size sz and returns a pointer to the array
+    template<class T> inline
+    T* allocate(const int sz)
+    {
+      T* array = new T[sz];
+      return array;
+    }
+
+    // allocates a memory of size ysz x xsz and returns a double pointer to it
+    template<class T> inline
+    T** allocate(const int ysz, const int xsz)
+    {
+      T** mat = new T*[ysz];
+      int i;
+
+      for(i=0; i<ysz; i++ )
+         mat[i] = new T[xsz];
+      // allocate<T>(xsz);
+
+      return mat;
+    }
+
+    // deallocates the memory and sets the pointer to null.
+    template<class T> inline
+    void deallocate(T* &array)
+    {
+      delete[] array;
+      array = NULL;
+    }
+
+    // deallocates the memory and sets the pointer to null.
+    template<class T> inline
+    void deallocate(T** &mat, int ysz)
+    {
+      if( mat == NULL ) return;
+
+      for(int i=0; i<ysz; i++)
+         deallocate(mat[i]);
+
+      delete[] mat;
+      mat = NULL;
+    }
+
+    // Converts the given polar coordinates of a point to cartesian
+    // ones.
+    template<class T1, class T2> inline
+    void polar2cartesian(T1 r, T1 t, T2 &y, T2 &x)
+    {
+      x = (T2)( r * cos( t ) );
+      y = (T2)( r * sin( t ) );
+    }
+
+
+    template<typename T> inline
+    void convolve_sym_( T* image, int h, int w, T* kernel, int ksize )
+    {
+      conv_horizontal( image, h, w, kernel, ksize );
+      conv_vertical  ( image, h, w, kernel, ksize );
+    }
+
+    template<class T>
+    inline void convolve_sym( T* image, int h, int w, T* kernel, int ksize, T* out=NULL )
+    {
+      if( out == NULL ) out = image;
+      else memcpy( out, image, sizeof(T)*h*w );
+      convolve_sym_(out, h, w, kernel, ksize);
+    }
+
+    // divides the elements of the array with num
+    template<class T1, class T2> inline
+    void divide(T1* arr, int sz, T2 num )
+    {
+      float inv_num = (float) (1.0 / num);
+
+      for( int i=0; i<sz; i++ )
+      {
+         arr[i] = (T1)(arr[i]*inv_num);
+      }
+    }
+
+    // returns an array filled with zeroes.
+    template<class T> inline
```

Comment:
[DONE] we have now Mat(r,c,CV_32F,Scalar(0)) or mat.setTo(Scalar(0)). I did not used std::vector in the code, Mat is much more flexible and can be casted /converted / re-dimed with much more ease.